### PR TITLE
Adding FGES

### PR DESCRIPTION
**Pull request by RobertGregg**
So it looks like I've got FGES working with the DiGraphs from `Graphs.jl`. I've also simplified some parts of the package I wrote:

- Got rid of the `message()` function as that was specific to running the package on a cluster
- Using good old `X \ y` to perform linear regression
- Changed the debug keyword with verbose to match the other algorithms

I added a test for `fges` (copying the test from pc2.jl) and it passed. Let me know if you want me to add other tests.

I also added two dependencies `Memoization` and `LRUCache`. FGES performs A LOT of linear regressions and often performs the same one more than once. These two packages basically keep a dictionary of inputs and outputs that it checks before running the `score` function. 

Let me know what you think!

### CODE REVIEW

File: src/fges.jl [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+==(a::T, b::T) where T <: ParseData = true
+
+
+####################################################################
+# Main Entry point for the Algorithm
+####################################################################
+
+"""
+    fges(data; penalty = 1.0, verbose=false)
+Compute a causal graph for the given observed data.
```
Comment (mschauer) [inline src/fges.jl:63]:
```suggestion

Compute a causal graph for the given observed data.
```

File: test/fges.jl [hunk 2]
```diff
@@ -0,0 +1,26 @@
+using Distributions, CausalInference, Graphs
+using Test, Random
+
+
+Random.seed!(100)
+N = 2000
+d = Normal()
+p = 0.01
```
Comment (mschauer) [inline test/fges.jl:8]:
```suggestion
```
unused?

File: src/fges_helper.jl [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+end
+calcNAyx(g, edge) = calcNAyx(g, dst(edge), src(edge))
+
+#for x-y, undirected neighbors of y not connected to x
+#calcT(g, y::Integer, x::Integer) = setdiff(neighbors_undirect(g,y), neighbors(g,x), x)
+function calcT(g, y, x)
+    neighborList = Int64[]
+
+    #loop through all vertices except for x
+    for vᵢ in Iterators.filter(!isequal(x),vertices(g))
```
Comment (mschauer) [inline src/fges_helper.jl:233]:
```suggestion
    for vᵢ in Iterators.filter(!isequal(x), vertices(g))
```

Comment (mschauer) [reply src/fges_helper.jl:233]: below too

File: src/fges_helper.jl [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+####################################################################
+# all* functions
+####################################################################
+
+#Generate an iterator equivalent to combinations(x,2) but produces a vector of tuples
+allpairs(v) = Iterators.filter(i -> isless(i...), Iterators.product(v,v))
+
+#Get every undirected edge in the entire graph g
+allundirected(g) = [Edge(nodePair) for nodePair in allpairs(vertices(g)) if isundirected(g,nodePair)]
```
Comment (mschauer) [inline src/fges_helper.jl:195]: Make it a generator?

File: src/fges_helper.jl [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+orientedge!(g, edge) = rem_edge!(g, reverse(edge))
+orientedge!(g, x, y) = rem_edge!(g, y, x)
+
+
+
+####################################################################
+# Neighborhood functions 
+####################################################################
+
+#Should these be non-allocating iterators?
```
Comment (mschauer) [inline src/fges_helper.jl:141]: Aren't they?

File: src/fges_helper.jl [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+end
+
+
+
+####################################################################
+# Modify Edges
+####################################################################
+
+"""
+    orientedge!(g, x, y)
```
Comment (mschauer) [inline src/fges_helper.jl:129]:
```suggestion
    orientedge!(g, x, y)
    
```

File: src/fges_helper.jl [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        end
+    end
+
+    return true
+end
+
+
+#Do the nodesRemoved block all semi-directed paths between src and dest?
+"""
+    isblocked(g, src, dest, nodesRemoved)
```
Comment (mschauer) [inline src/fges_helper.jl:79]:
```suggestion
    isblocked(g, src, dest, nodesRemoved)
    
```

**Comment by mschauer** [comment]
Very nice, no problem with the additional dependencies. One could think of getting rid of camel case in function names `findNextEquivClass!` => `find_next_equiv_class!` but that's a bit busy-work. I left some comments!

File: src/fges.jl [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+
+        #Ensure that the penalty is the same type as the data (e.g. Float32)
+        penalty = baseType(penalty)
+
+        return new{M, baseType}(data, normAugData, numFeatures, numObservations, penalty)
+    end
+end  
+
+#Simple structure to hold the current edge, a subset of neighbors, and a score change
+Base.@kwdef mutable struct Step{A,B}
```
Comment (mschauer) [inline src/fges.jl:37]:
```suggestion
Base.@kwdef mutable struct FESStep{A,B}
```
This should have a more specific name

File: src/fges.jl [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+    edge::Edge = Edge(1,2)
```
Comment (mschauer) [inline src/fges.jl:38]: Could be that this should be `SimpleEdge` as Edge is an abstract type. Is this struct performance critical / used in a tight loop?

File: src/fges.jl [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+#This could be dangerous. The @memoize macro has to check if ParseData is the same argument. We only ever define one immutable ParseData, so it will never change. Any equality check should be true.
+==(a::T, b::T) where T <: ParseData = true
+
+
+####################################################################
+# Main Entry point for the Algorithm
+####################################################################
+
+"""
+    fges(data; penalty = 1.0, verbose=false)
```
Comment (mschauer) [inline src/fges.jl:62]:
```suggestion
    fges(data; penalty = 1.0, verbose=false)
    
```
I am adding these spaces after the function signature in the doc string

**Comment by mschauer** [comment]
You committed the manifest and a settings file by accident, I think this might cause the test failure.

**Comment by mschauer** [comment]
What are your thoughts on this?

**Comment by RobertGregg** [comment]
Ah apologies, I completely dropped the ball on this. I'll look at your comments this weekend in more detail, but overall they seem like good/minor additions. I'll make sure the manifest file doesn't get committed and check out the `Edge()` struct. That might have been leftover from when I was converting to using Graphs.jl

File: src/fges_helper.jl (hunk 5 above)
Comment (RobertGregg) [reply src/fges_helper.jl:141]: I think they weren't originally, but I changed it and forgot to remove the comment.

File: src/fges.jl (hunk 9 above)
Comment (RobertGregg) [reply src/fges.jl:38]: Very good point. We should change this to a concrete type

File: src/fges_helper.jl (hunk 4 above)
Comment (RobertGregg) [reply src/fges_helper.jl:195]: Do you mean something like this?
```suggestion
allundirected(g) = (Edge(nodePair) for nodePair in allpairs(vertices(g)) if isundirected(g,nodePair))
```

**Review by RobertGregg — COMMENTED** [review]
All the suggestions seem good. I've only done a pull request once before so sorry if anything looks weird. I'll try and commit the suggestions and remove the manifest file to see if things check out.

File: src/fges_helper.jl (hunk 4 above)
Comment (mschauer) [reply src/fges_helper.jl:195]: Yes.

**Comment by mschauer** [comment]
I think you forgot to add the new test to the runtest file.

**Comment by mschauer** [comment]
Ps: If it is easier, just make a new PR.

**Comment by mschauer** [comment]
Continued in https://github.com/mschauer/CausalInference.jl/pull/92

**Pull request closed without merging**